backend/src/impl/db_utils/dataset_db_utils.py

```diff
@@ -1,42 +1,125 @@
+        self.metadatas: list[DatasetMetadata] = []
+        for dataset_name, v_dataset in data.items():
+            for sub_dataset, v_sub in v_dataset["sub_datasets"].items():
+                i = len(self.metadatas)
+                # Names
+                if dataset_name not in self.name_dict:
+                    self.name_dict[dataset_name] = []
+                self.name_dict[dataset_name].append(i)
+                # Ids
+                dataset_id = f"{dataset_name}---{sub_dataset}"
+                self.id_dict[dataset_id] = i
+                # Tasks
+                tasks = v_dataset.get("task_templates", None)
+                if tasks is None:
+                    tasks = []
+                else:
+                    for task in tasks:
+                        if task not in self.task_dict:
+                            self.task_dict[task] = []
+                        self.task_dict[task].append(i)
+                # Create document
+                doc = {
+                    "dataset_id": dataset_id,
+                    "dataset_name": dataset_name,
+                    "sub_dataset": None if sub_dataset == "__NONE__" else sub_dataset,
+                    "split": v_sub["splits"],
+                    "tasks": tasks,
+                }
+                self.metadatas.append(DatasetMetadata.from_dict(doc))
+        self.name_trie = marisa_trie.Trie(self.name_dict.keys())
+
+
 class DatasetDBUtils:
+
+    online_path = "https://raw.githubusercontent.com/ExpressAI/DataLab/main/utils/dataset_info.jsonl"  # noqa
+    _cached_db: DatasetDB | None = None
+    _cached_time: datetime | None = None
+    _cached_lifetime: timedelta = timedelta(hours=6)
+
+    @staticmethod
+    def get_dataset_db() -> DatasetDB:
+        """
+        Get a collection of datasets, indexed by dataset name, containing every
+        sub_dataset
+        """
+        if DatasetDBUtils._cached_db is None or (
+            datetime.now() - unwrap(DatasetDBUtils._cached_time)
+            > unwrap(DatasetDBUtils._cached_lifetime)
+        ):
+            local_path = cache_online_file(
+                DatasetDBUtils.online_path,
+                "info/dataset_info.jsonl",
+                lifetime=DatasetDBUtils._cached_lifetime,
+            )
+            datasets = {}
+            with open(local_path, "r") as fin:
+                for line in fin:
+                    data = json.loads(line)
+                    for k, v in data.items():
+                        # skip 'ERROR' or 'SKIPPED' entries
+                        if isinstance(v, dict) and len(v) > 0:
+                            datasets[k] = v
+            DatasetDBUtils._cached_db = DatasetDB(datasets)
+            DatasetDBUtils._cached_time = datetime.now()
+        return DatasetDBUtils._cached_db
+
+    @staticmethod
+    def find_dataset_by_id(dataset_id: str) -> DatasetMetadata | None:
+        my_db = DatasetDBUtils.get_dataset_db()
+        _id = my_db.id_dict.get(dataset_id)
```

Comment:
`_id` is identical to the `i` in `DatasetDB`'s  `__init__` function. We can standardize the variable name.